### PR TITLE
Geometric surface for sharding

### PR DESCRIPTION
**Pull request by jklamer**
Use GeometricSurface to simplify and expand sharding code.

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/geography/sharding/Sharding.java [hunk 1]
```diff
@@ -81,39 +76,11 @@ static Sharding forString(final String sharding)
     /**
      * Generate shards. This needs to be deterministic!
      *
-     * @param multiPolygon
+     * @param surface
      *            The bounds to limit the shards.
      * @return The shards {@link Iterable}.
      */
-    default Iterable<? extends Shard> shards(final MultiPolygon multiPolygon)
```
Comment (jwpgage) [inline src/main/java/org/openstreetmap/atlas/geography/sharding/Sharding.java]: Why is there no longer a default?

File: src/test/java/org/openstreetmap/atlas/geography/sharding/SlippyTileTest.java [hunk 2]
```diff
@@ -66,9 +68,9 @@ public void testMultiPolygonShards()
                 .wkt("MULTIPOLYGON (((40 40, 20 45, 45 30, 40 40)),"
                         + "((20 35, 10 30, 10 10, 30 5, 45 20, 20 35),"
                         + "(30 20, 20 15, 20 25, 30 20)))");
-        @SuppressWarnings("unchecked")
-        final Set<Shard> coveredShards = (Set<Shard>) new SlippyTileSharding(8)
-                .shards(multiPolygon);
+        final Set<Shard> coveredShards = StreamSupport
+                .stream(new SlippyTileSharding(8).shards(multiPolygon).spliterator(), false)
```
Comment (jwpgage) [inline src/test/java/org/openstreetmap/atlas/geography/sharding/SlippyTileTest.java]: Did you notice a performance increase using `spliterator()`?

Comment (jklamer) [reply src/test/java/org/openstreetmap/atlas/geography/sharding/SlippyTileTest.java]: Not sure. This was just implement set construction in one line. The previous test used the knowledge that the returned iterable was a Set and directly cast to that. That broke when I got rid of the default.

File: src/main/java/org/openstreetmap/atlas/geography/sharding/Sharding.java (hunk 1 above)
Comment (jklamer) [reply src/main/java/org/openstreetmap/atlas/geography/sharding/Sharding.java]: Both Polygon and MultiPolygon are GeometricSurfaces. That interface contains the contract for the overlapping logic that is in this default function. Ideally, all MultiPolygon intersection is handled by the shape class and no implementation knowledge is needed. Deleting this removes duplicate work and prevents it breaking if MultiPolygon implementation is ever changed.

**Pull request merged**